### ISSUE TITLE
Proposal: zig test run single test case

### ISSUE THREAD

**Issue by star-tek-mb** [ziglang/zig#13524]
Zig version

`0.11.0-dev.109+a2e67173d`

As today, if we have a source:
```zig
test "read" {
...
}

test "read struct" {
...
}
```

We can't run single test case `read` with `zig test file.zig --test-filter "read"`. It will run both tests. There is no way to run specific test.

# Proposal

### Option 1:

`--test-filter` will use pattern matching (glob or wildcard will be sufficent). For now it uses `std.mem.indexOf`. It would be breaking change maybe (`--test-filter` not used in compiler tests i think, only for user code). And will require to write a pattern matching code in stdlib.

### Option 2:

Add another option to specify exact test case to run: `--test-name` or `--test-case`. It would be non breaking and easy to achieve.

Any ideas welcome.

### Option 3

Write simple wildcard matching in src/test.zig, or use [reference code](https://github.com/mattnite/glob/blob/main/src/main.zig#L55). I prefer this option, i can even try open pull request for that.

**Comment by thomas-mangin** [comment]
Looking for this feature, I found this issue, just to notice that it is already present (and working) using HEAD. Does this issue just need closing?

```bash
❯ zig version
0.11.0-dev.1580+a5b34a61a
```

```bash
❯ zig test --help | grep -A2 "Test Options"
Test Options:
  --test-filter [text]           Skip tests that do not match filter
  --test-name-prefix [text]      Add prefix to all tests
```

**Comment by star-tek-mb** [comment]
@thomas-mangin 

> We can't run single test case `read` with `zig test file.zig --test-filter "read"`. It will run both tests. There is no way to run specific test.

It's complicated by the way. Because you can use any literal in test's names (not just string literals).

**Comment by Vexu** [comment]
> Looking for this feature, I found this issue, just to notice that it is already present (and working) using HEAD. Does this issue just need closing?

This is a proposal to enhance that feature.

**Comment by icholy** [comment]
The `--test-filter` flag could be modified to support a `$` suffix which behaves in the same way a regex does. For example:

``` zig
test "foo" {}
test "foobar" {}
```

* `--filter-test foo` would run both tests.
* `--filter-test 'foo$'` would only run the first test.

``` diff
diff --git a/test/src/CompareOutput.zig b/test/src/CompareOutput.zig
index 58c1238705..6abc10aa62 100644
--- a/test/src/CompareOutput.zig
+++ b/test/src/CompareOutput.zig
@@ -93,7 +93,11 @@ pub fn addCase(self: *CompareOutput, case: TestCase) void {
                 case.name,
             }) catch @panic("OOM");
             for (self.test_filters) |test_filter| {
-                if (mem.indexOf(u8, annotated_case_name, test_filter)) |_| break;
+                if (mem.endsWith(u8, test_filter, "$")) {
+                    if (mem.endsWith(u8, annotated_case_name, test_filter[0..test_filter.len-1])) |_| break;
+                } else {
+                    if (mem.indexOf(u8, annotated_case_name, test_filter)) |_| break;
+                }
             } else if (self.test_filters.len > 0) return;

             const exe = b.addExecutable(.{
```

**Comment by icholy** [comment]
Alternatively, there could be a `@` prefix which must be used with the FQN. For example:

```
--test-filter @main.test.foo
```

**Comment by dweiller** [comment]
I don't think it makes sense to extend the `--test-filter` CLI parameter to allow specifying a particular test name or FQN as there isn't any syntax that will work without escaping since decls names using `@"..."` syntax will cause issues/ambiguity. I think it would be better to just have other CLI parameters for filtering in different ways, like a `--test-name` and/or `--test-fqn` for specifying the name of the test or FQN respectively (here by the 'name' of a test I mean the last component of the FQN).

**Comment by nektro** [comment]
it would be easy for tooling to be able to reference them by source location eg `./path:line:col`

**Comment by icholy** [comment]
@nektro is the `col` necessary?

**Comment by nektro** [comment]
far, in practice probably not; it would almost always be zero/4/8/etc

**Comment by dweiller** [comment]
You can write multiple test blocks on a single line (not that I think a human would choose to do so):
```zig
test "a" {} test "b" {}
```
So if you wanted to select tests by source location you would need to column.

**Comment by andrewrk** [comment]
Reconsidering in light of #25352. I think IDEs want this feature too.

**Comment by andrewrk** [comment]
Here's what's accepted:

Additional `--test-filter-exact [name]` CLI argument. It can appear any number of times, creating a set of unit tests to run. When 1 or more of these arguments appear, `--test-filter [name]` arguments are not allowed. If any named unit test is not found, Zig prints an error message to stderr and exits with failure code. Only exactly the set of named unit tests (fully qualified names) are run.

**Comment by nektro** [comment]
happy to see this open again :)

i suggest `--test-filter-exact [file:line]` as it would be more precise and also support unnamed tests, particularly handy for the editor and tooling use case mentioned above

**Comment by dweiller** [comment]
> Here's what's accepted:
> 
> Additional `--test-filter-exact [name]` CLI argument. It can appear any number of times, creating a set of unit tests to run. When 1 or more of these arguments appear, `--test-filter [name]` arguments are not allowed. If any named unit test is not found, Zig prints an error message to stderr and exits with failure code. Only exactly the set of named unit tests (fully qualified names) are run.

At the moment, it is possible for multiple tests to have the same fully qualified name; it's unclear to me whether this is a bug[^1]. An example file that has tests with clashing FQNs is the following:
```zig
test "test.name" {}

const @"test" = struct {
    test "name" {}
};

comptime {
    _ = @"test";
}
```

I think it could be a bit confusing to get multiple tests matching with a single `--test-filter-exact` for cases like the above, as I imagine users would expect a FQN to be unique (given that I think you have to use `@"..."` syntax to get clashing FQNs, maybe this is a bad assumption on my part) and there could be situations where you effectively get a false positive result if the intended test isn't actually referenced. If the ability for test FQNs to clash is intended is the plan to simply include all matching tests and assume that the user knows what they're doing?

[^1]:  I haven't been able to cause a compiler crash/issue with clashing test FQNs, but some parts of the codebase look like they expect or want FQNs to be distinct, but I don't have enough context to be sure.
